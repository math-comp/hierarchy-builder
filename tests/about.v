From Coq Require Import ZArith ssrfun ssreflect.
From HB Require Import structures.
From HB Require Import demo1.hierarchy_5.

HB.instance
Definition _ :=
  AddAG_of_TYPE.Build Z 0%Z Z.add Z.opp
    Z.add_assoc Z.add_comm Z.add_0_l Z.add_opp_diag_l.

HB.instance
Definition _ :=
  Ring_of_TYPE.Build Z 0%Z 1%Z Z.add Z.opp Z.mul
    Z.add_assoc Z.add_comm Z.add_0_l Z.add_opp_diag_l
    Z.mul_assoc Z.mul_1_l Z.mul_1_r
    Z.mul_add_distr_r Z.mul_add_distr_l.

(* mixin *)
HB.about AddMonoid_of_TYPE.

(* mixin constructor *)
HB.about AddMonoid_of_TYPE.Build.

(* structure *)
HB.about AddAG.type.

(* class *)
HB.about AddMonoid.

(* factory *)
HB.about Ring_of_AddAG.

(* factory constructor *)
HB.about Ring_of_AddAG.Build.

(* operation *)
HB.about add.

(* canonical proj/value *)
HB.about AddAG.sort.

(* canonical value *)
HB.about Z.

(* coercion *)
HB.about hierarchy_5_Ring_class__to__hierarchy_5_SemiRing_class.
HB.about hierarchy_5_Ring__to__hierarchy_5_SemiRing.

(* builder *)
HB.about Builders_40.hierarchy_5_Ring_of_AddAG__to__hierarchy_5_BiNearRing_of_AddMonoid.

HB.locate BinNums_Z__canonical__hierarchy_5_AddAG.

(* Test minimally qualified names *)
Module Import hierarchy_5.
Module AddComoid.
End AddComoid.
End hierarchy_5.
HB.about Z.
