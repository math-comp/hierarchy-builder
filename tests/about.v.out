HB: AddMonoid_of_TYPE is a factory
    (from "./examples/demo1/hierarchy_5.v", line 10)
HB: AddMonoid_of_TYPE operations and axioms are:
    - zero
    - add
    - addrA
    - add0r
    - addr0
HB: AddMonoid_of_TYPE requires the following mixins:
HB: AddMonoid_of_TYPE provides the following mixins:
    - AddMonoid_of_TYPE

HB: AddMonoid_of_TYPE.Build is a factory constructor
    (from "./examples/demo1/hierarchy_5.v", line 10)
HB: AddMonoid_of_TYPE.Build requires its subject to be already equipped with:
HB: AddMonoid_of_TYPE.Build provides the following mixins:
    - AddMonoid_of_TYPE
HB: arguments: AddMonoid_of_TYPE.Build S zero add addrA add0r addr0
    - S : Type
    - zero : AddMonoid.sort S
    - add : S -> S -> S
    - addrA : associative add
    - add0r : left_id 0%G add
    - addr0 : right_id 0%G add

HB: AddAG.type is a structure (from "./examples/demo1/hierarchy_5.v", line 73)
HB: AddAG.type characterizing operations and axioms are:
    - addNr
    - opp
HB: AddAG is a factory for the following mixins:
    - AddMonoid_of_TYPE
    - AddComoid_of_AddMonoid
    - AddAG_of_AddComoid (* new, not from inheritance *)
HB: AddAG inherits from:
    - AddMonoid
    - AddComoid
HB: AddAG is inherited by:
    - Ring

HB: AddMonoid.type is a structure
    (from "./examples/demo1/hierarchy_5.v", line 17)
HB: AddMonoid.type characterizing operations and axioms are:
    - addr0
    - add0r
    - addrA
    - add
    - zero
HB: AddMonoid is a factory for the following mixins:
    - AddMonoid_of_TYPE (* new, not from inheritance *)
HB: AddMonoid inherits from:
HB: AddMonoid is inherited by:
    - AddComoid
    - AddAG
    - BiNearRing
    - SemiRing
    - Ring

HB: Ring_of_AddAG is a factory
    (from "./examples/demo1/hierarchy_5.v", line 108)
HB: Ring_of_AddAG operations and axioms are:
    - one
    - mul
    - mulrA
    - mulr1
    - mul1r
    - mulrDl
    - mulrDr
HB: Ring_of_AddAG requires the following mixins:
    - AddMonoid_of_TYPE
    - AddComoid_of_AddMonoid
    - AddAG_of_AddComoid
HB: Ring_of_AddAG provides the following mixins:
    - BiNearRing_of_AddMonoid
Doc: Builds a Ring from an Abelian Group: the absorbing properties mul0r and
     mul0r are derived from addrC and the other ring axioms, following a proof
     of Hankel (Gerhard Betsch. On the beginnings and development of near-ring
     theory. In Near-rings and near-fields. Proceedings of the conference held
     in Fredericton, New Brunswick, July 18-24, 1993, pages 1–11. Mathematics
     and its Applications, 336. Kluwer Academic Publishers Group, Dordrecht,
     1995).
HB: Ring_of_AddAG.Build is a factory constructor
    (from "./examples/demo1/hierarchy_5.v", line 108)
HB: Ring_of_AddAG.Build requires its subject to be already equipped with:
    - AddMonoid_of_TYPE
    - AddComoid_of_AddMonoid
    - AddAG_of_AddComoid
HB: Ring_of_AddAG.Build provides the following mixins:
    - BiNearRing_of_AddMonoid
HB: arguments: Ring_of_AddAG.Build A [one] [mul] mulrA mulr1 mul1r mulrDl mulrDr
    - A : Type
    - one : A
    - mul : A -> A -> A
    - mulrA : associative mul
    - mulr1 : left_id one mul
    - mul1r : right_id one mul
    - mulrDl : left_distributive mul add
    - mulrDr : right_distributive mul add
Doc: Builds a Ring from an Abelian Group: the absorbing properties mul0r and
     mul0r are derived from addrC and the other ring axioms, following a proof
     of Hankel (Gerhard Betsch. On the beginnings and development of near-ring
     theory. In Near-rings and near-fields. Proceedings of the conference held
     in Fredericton, New Brunswick, July 18-24, 1993, pages 1–11. Mathematics
     and its Applications, 336. Kluwer Academic Publishers Group, Dordrecht,
     1995).
HB: add is an operation of structure AddMonoid
    (from "./examples/demo1/hierarchy_5.v", line 17)
HB: add comes from mixin AddMonoid_of_TYPE
    (from "./examples/demo1/hierarchy_5.v", line 10)
HB: AddAG.sort is a canonical projection
    (from "./examples/demo1/hierarchy_5.v", line 73)
HB: AddAG.sort has the following canonical values:
    - @eta 
    - Ring.sort (from "./examples/demo1/hierarchy_5.v", line 196)
    - Z 
HB: AddAG.sort is a coercion from AddAG to Sortclass
    (from "./examples/demo1/hierarchy_5.v", line 73)
HB: Z is canonically equipped with structures:
    - AddMonoid
      AddComoid
      AddAG
      (from this file, line 5)
    - BiNearRing
      SemiRing
      Ring
      (from this file, line 10)
HB: hierarchy_5_Ring_class__to__hierarchy_5_SemiRing_class is a coercion from
    Ring to SemiRing (from "./examples/demo1/hierarchy_5.v", line 196)
HB: hierarchy_5_Ring__to__hierarchy_5_SemiRing is a coercion from
    Ring to SemiRing (from "./examples/demo1/hierarchy_5.v", line 196)
HB: todo HB.about for builder from Ring_of_AddAG to BiNearRing_of_AddMonoid
HB: synthesized in file File "(stdin)", line 5, column 0, character 127:
Interactive Module hierarchy_5 started
Interactive Module AddComoid started
HB: Z is canonically equipped with structures:
    - AddMonoid
      demo1.hierarchy_5.AddComoid
      AddAG
      (from this file, line 5)
    - BiNearRing
      SemiRing
      Ring
      (from this file, line 10)
